src/Service/VendorService/OverDrive/Api/Client.php

```diff
@@ -0,0 +1,316 @@
+<?php
+/**
+ * @file
+ * OverDrive API client
+ *
+ * @see https://developer.overdrive.com/discovery-apis
+ */
+
+namespace App\Service\VendorService\OverDrive\Api;
+
+use App\Service\VendorService\OverDrive\Api\Exception\AuthException;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\GuzzleException;
+use Psr\Cache\InvalidArgumentException;
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+
+/**
+ * Class Client.
+ */
+class Client
+{
+    private const MAX_RETRIES = 5;
+
+    private const OAUTH_URL = 'https://oauth.overdrive.com/token';
+    private const USER_AGENT = 'cover.dandigbib.org';
+
+    /** @var string */
+    private $libraryAccountEndpoint;
+
+    /** @var string */
+    private $clientId;
+
+    /** @var string */
+    private $clientSecret;
+
+    /** @var AdapterInterface */
+    private $cache;
+
+    /** @var ClientInterface */
+    private $httpClient;
+
+    /** @var string */
+    private $authorization;
+
+    /** @var string */
+    private $productsEndpoint;
+
+    /**
+     * Client constructor.
+     *
+     * @param AdapterInterface $cache
+     * @param ClientInterface $httpClient
+     */
+    public function __construct(AdapterInterface $cache, ClientInterface $httpClient)
+    {
+        $this->cache = $cache;
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * Set OverDrive authentication credentials.
+     *
+     * @see https://developer.overdrive.com/apis/client-auth
+     * @see https://developer.overdrive.com/apis/library-account
+     *
+     * @param string $libraryAccountEndpoint
+     *   The custom endpoint for the account used
+     * @param string $clientId
+     *   Client id
+     * @param string $clientSecret
+     *   Client secret
+     */
+    public function setCredentials(string $libraryAccountEndpoint, string $clientId, string $clientSecret): void
+    {
+        $this->libraryAccountEndpoint = $libraryAccountEndpoint;
+        $this->clientId = $clientId;
+        $this->clientSecret = $clientSecret;
+    }
+
+    /**
+     * Get cover url from OverDrive crossRefId.
+     *
+     * @param string $crossRefId
+     *   The OverDrive 'crossRefId'
+     *
+     * @return string|null
+     *   The cover url or null
+     *
+     * @throws GuzzleException
+     * @throws AuthException
+     * @throws InvalidArgumentException
+     */
+    public function getCoverUrl(string $crossRefId): ?string
+    {
+        $requests = 0;
+
+        // The OverDrive api fails at random but still gives a 200 response.
+        // Only indication of failure is that the 'products' key is missing
+        // from the response body.
+        // If the products key is missing we retry the request 'MAX_RETRIES' times
```

Comment:
This sounds a lot as "The Retry Pattern" or maybe the "Circuit Breaker pattern" (https://martinfowler.com/bliki/CircuitBreaker.html) that I have talk about before with https://github.com/ackintosh/ganesha/ with retry strategies.

Reply:
Don't think so. 
- We are not running in parallel
- We don't need to stop on error. Just retry
- Errors are not proper errors the http client can detect (`200` is returned)